---
title: "Report: `r params$project`"
output:
  rmdformats::robobook:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: false
    highlight: tango
params:
  project: test-gwas
  date: 2021-08-09
  version: v0.1.4
  regenie_merged: ../output/test-gwas/results/Y1.regenie.gz
  regenie_filename: test-gwas.Y1.regenie.all.gz
  phenotype: Y1
  covariates:
  phenotype_file: ../tests/input/phenotype.txt
  regenie_step1_log: NO_LOG
  manhattan_plot_ylimit: 2
  regenie_step2_log: ../output/test-gwas/logs/test-gwas.step2.log
  annotated_tophits_filename: ../output/test-gwas/tophits/Y1.regenie.filtered.tophits.annotated.txt.gz
---

```{r setup, include=FALSE}
library(qqman)
library(DT)
library(data.table)
library(R.utils)
library(ggplot2)
library(skimr)
library(kableExtra)
library(ramwas)
library(dplyr)
```

## Project Summary

| Parameter     | Value                                      |
| ------------- |--------------------------------------------|
| Project       | `r params$project`                         |
| Pipeline Version       | `r params$version`                |
| Date       | `r params$date`                         |
| Phenotype File      | `r params$phenotype_file`            |
| Phenotype       | `r params$phenotype`                     |
| Covariates       | `r params$covariates`                     |
| Regenie Output      | `r params$regenie_filename`             |

## Regenie Log

### Step 1

```{r, echo=FALSE, results='asis'}
if(file.exists(params$regenie_step1_log)) {
step1_log <- read.table(params$regenie_step1_log,sep ="\t", header = TRUE, dec =".")
kable(step1_log)
} else {
    cat("*No log file available, since regenie step 1 is skipped.*")
}
```

### Step 2

```{r, echo=FALSE}
step2_log <- read.table(params$regenie_step2_log,
                      sep ="\t", header = TRUE, dec =".")
kable(step2_log)
```

## Phenotype Statistics

### Overview

```{r, echo=FALSE}
phenotypeTable <- read.table(params$phenotype_file, header=TRUE, sep=" ", dec = ".")

kable(skim(phenotypeTable) %>%
  dplyr::filter(skim_variable == params$phenotype)) %>%
  kableExtra::scroll_box(width = "100%")
```

### Histogram
```{r, echo=FALSE}
ggplot(phenotypeTable, aes_string(x=params$phenotype)) +
  geom_histogram(color="black", fill="white",bins=30)
```

## Manhattan Plot

```{r, echo=FALSE, results='asis'}
regenieTable <- fread(params$regenie_merged, select = c("CHROM","GENPOS","LOG10P","ID"), header=TRUE,tmpdir="$PWD")

if(params$manhattan_plot_ylimit == 0) {
  manhattan(regenieTable, chr="CHROM", bp="GENPOS", p="LOG10P", snp="ID", logp= FALSE, cex = 0.3, col = c("blue4", "orange3"))
} else {
  manhattan(regenieTable, chr="CHROM", bp="GENPOS", p="LOG10P", snp="ID", logp= FALSE, cex = 0.3, col = c("blue4", "orange3"), ylim = c(0, params$manhattan_plot_ylimit))
  cat(paste("Please note that the Y-axis limit has been set to '", params$manhattan_plot_ylimit,"'. Please use the 'manhattan_plot_ylimit' parameter to adapt this behaviour."))
}
```

## QQ Plot

```{r, echo=FALSE}
qqPrepare <- qqPlotPrepare(pvalues=regenieTable$LOG10P,ismlog10 = T)
qqPlotFast(qqPrepare)

rm(qqPrepare)
rm(regenieTable)
```

## Top Genes

```{r, echo=FALSE, message=FALSE, warning=FALSE}

annotatedTopHits <- fread(params$annotated_tophits_filename, select = c("CHROM", "GENPOS", "LOG10P", "ID", "GENE_NAME", "GENE_CHROMOSOME", "GENE_START", "GENE_END", "DISTANCE"), header=TRUE)

topHitsGroupedByGene <-
  annotatedTopHits %>%
  group_by(GENE_NAME, GENE_CHROMOSOME, GENE_START, GENE_END) %>%
  summarise(GENE_LOG10P = max(LOG10P), VARIANTS = n()) %>%
  arrange(desc(GENE_LOG10P))

datatable(topHitsGroupedByGene)

```

---

<small>
This report has been created with **[GWAS-Regenie `r params$version`](https://github.com/genepi/gwas-regenie)**, a Nextflow Pipeline developed by [Lukas Forer](mailto:lukas.forer@i-med.ac.at) and [Sebastian Sch√∂nherr](mailto:sebatian.schoenherr@i-med.ac.at), Institute of Genetic Epidemiology, Medical University of Innsbruck.
</small>
