---
title: "Report: `r params$project`"
output:
  rmdformats::robobook:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: false
    highlight: tango
params:
  project: test-gwas
  date: 2021-08-09
  version: v0.1.9
  regenie_merged: ../output/test-gwas/results/Y1.regenie.gz
  regenie_filename: test-gwas.Y1.regenie.all.gz
  phenotype: Y1
  covariates:
  phenotype_file: ../output/test-gwas/validated_input/phenotype.validated.txt
  regenie_step1_log: ../output/test-gwas/logs/test-gwas.step1.log
  regenie_step2_log: ../output/test-gwas/logs/test-gwas.step2.log
  phenotype_log: ../output/test-gwas/logs/phenotype.validated.log
  covariate_log: NO_COV_LOG
  plot_ylimit: 0
  manhattan_annotation_enabled: true
  annotation_min_log10p: 5
  annotated_tophits_filename: ../output/test-gwas/results/tophits/Y1.regenie.filtered.annotated.txt.gz
---

```{r setup, include=FALSE}
library(qqman)
library(DT)
library(data.table)
library(R.utils)
library(ggplot2)
library(skimr)
library(kableExtra)
library(ramwas)
library(dplyr)
```

## Project Summary

| Parameter     | Value                                      |
| ------------- |--------------------------------------------|
| Project       | `r params$project`                         |
| Pipeline Version       | `r params$version`                |
| Date       | `r params$date`                         |
| Phenotype File      | `r params$phenotype_file`            |
| Phenotype       | `r params$phenotype`                     |
| Covariates       | `r params$covariates`                     |
| Regenie Output      | `r params$regenie_filename`             |

## Phenotype Statistics

### Overview

```{r, echo=FALSE}
phenotypeTable <- read.table(params$phenotype_file, header=TRUE, sep="\t", dec = ".")
kable(skim(phenotypeTable) %>%
  dplyr::filter(skim_variable == params$phenotype)) %>%
  kableExtra::scroll_box(width = "100%")
```

### Histogram
```{r, echo=FALSE}
ggplot(phenotypeTable, aes_string(x=params$phenotype)) +
  geom_histogram(color="black", fill="white",bins=30)
```

## Manhattan Plot

```{r, echo=FALSE, message=FALSE, warning=FALSE, dpi=300}
#sources: https://www.r-graph-gallery.com/101_Manhattan_plot.html with modifications

regenieTable <- fread(params$regenie_merged, select = c("CHROM","GENPOS","LOG10P","ID", "A1FREQ"), header=TRUE, tmpdir="$PWD")

#source: https://www.r-graph-gallery.com/101_Manhattan_plot.html with modifications

annotatedTopHits <- fread(params$annotated_tophits_filename, select = c("ID", "GENE_NAME"), header=TRUE, tmpdir="$PWD")
#merge gene names of tophits to regenieTable

annotatedTopHits <- annotatedTopHits %>%
    mutate(ID = as.character(ID))
regenieTable <- regenieTable %>%
    mutate(ID = as.character(ID))

regenieTable_annotated <- left_join(regenieTable, annotatedTopHits, by = "ID")

#prepare for plotting
don <- regenieTable_annotated %>%

  # Compute chromosome size
  group_by(CHROM) %>%
  summarise(CHROM_len=max(GENPOS)) %>%

  # Calculate cumulative start position of each chromosome
  mutate(tot=cumsum(as.numeric(CHROM_len))-CHROM_len) %>%
  select(-CHROM_len) %>%

  # Add this info to the initial dataset
  left_join(regenieTable_annotated, ., by=c("CHROM"="CHROM")) %>%

  # Add a cumulative position of each SNP
  arrange(CHROM, GENPOS) %>%
  mutate( GENPOScum=GENPOS+tot)

rm(regenieTable_annotated)

# Prepare X axis
axisdf = don %>% group_by(CHROM) %>% summarize(center=( max(GENPOScum) + min(GENPOScum) ) / 2 )

p <- ggplot(don, aes(x=GENPOScum, y=LOG10P)) +

    # Show all points
    geom_point(aes(color=as.factor(CHROM))) +
    scale_color_manual(values = rep(c("#779ECB", "#03254c"), 22 )) +

    # custom X axis:
    scale_x_continuous(label = axisdf$CHROM,
                       breaks= axisdf$center,
                       name = "Chromosome",
                       expand = c(0,0),
                       guide = guide_axis(check.overlap = TRUE)) +
  # custom y step 1annotation_limit
    scale_y_continuous(expand = c(0, 0),
                       name=expression(-log[10](italic(p))),
                       limits = c(0, sum(max(regenieTable$LOG10P), 0.5))) +
  geom_hline(yintercept = -log10(5e-08),
             linetype ="longdash",
             color ="firebrick") + # genomewide significance
  geom_hline(yintercept = -log10(1e-5),
             linetype ="longdash",
             color ="darkgrey")  + # suggestive significance

    # Custom the theme:
    theme_classic() +
    theme(
      legend.position="none",
      panel.border = element_blank(),
      axis.text = element_text(size = 12,
                               color = "black"),
      axis.title = element_text(size = 14),
      axis.ticks = element_line(color = "black")
      )

# custom y axis step 2 (depends if ymax is defined)

 if(params$plot_ylimit == 0) {
  p2 <- p + scale_y_continuous(expand = c(0, 0),
                       name=expression(-log[10](italic(P))),
                       limits = c(0, sum(max(regenieTable$LOG10P), 0.5)))
} else {
  p2 <- p + scale_y_continuous(expand = c(0, 0),
                       name=expression(-log[10](italic(P))),
                       limits = c(0, params$plot_ylimit))
}


 if(!params$manhattan_annotation_enabled) {
  p2
} else {
  p2 + ggrepel::geom_text_repel(data = don %>% group_by(GENE_NAME) %>%
                             top_n(1, LOG10P) %>%
                               slice(1) %>%
                             filter(LOG10P > params$annotation_min_log10p),
                            aes(label=GENE_NAME), color = 1)
}


```

## QQ Plot

```{r, echo=FALSE, results='asis'}
compute.inflation <- function(x, log=FALSE) {
  if (sum(is.na(x)) > 1) {
    x <- x[!is.na(x)]
  }

  if (log){
    x <- 10**(-x)
  }

  ## lambda chisq <- qchisq(1-x,1)
  chisq <- qchisq(x, 1, lower.tail = FALSE)
  lambda <- median(chisq) / qchisq(0.5, 1)

  return(lambda)
}

#---- Create null distribution for qqplot ----
create_nulldist <- function(x, log=FALSE){
  n <- length(x)
  expected <- -log10(1:n / n)
  if (log){
    xl <- x
  } else {
    xl <- -log10(x)
  }
  ix <- order(xl, decreasing=TRUE)
  newexp <- rep(1, n)
  newexp[ix] <- expected
  return(newexp)
}


myqqplot_mafbreak <- function(
  df,
  mafbreaks=c(0, 0.0005, 0.005, 0.05, 1),
  mafcol = "AF_Allele2",
  pvalcol = "p.value",
  log=FALSE,
  title="QQPLOT"
){
  mylog10 <- function(x){
    return(-log10(x))
  }

  df$MAFBRK <- cut(df %>% pull({{ mafcol }}), breaks=mafbreaks)


  df <- df %>%
    group_by(MAFBRK)  %>%
    mutate(
      across(
        any_of(pvalcol),
        ~create_nulldist(.x, log=log),
        .names="EXP")
    )
  overall.lmb <- compute.inflation(df %>% pull({{ pvalcol }}), log=log)

  # mylambdas <- df %>% group_by(MAFBRK) %>%
  #   summarise(lambda = compute.inflation(-log10(p.value))$lambda,
  #             EXP=0, p.value=-log10(min(p.value)))

  if (!log){
    df <- df %>% mutate_at(vars({{pvalcol}}), ~mylog10)
  }

  qqp <- ggplot(df, aes_string("EXP", pvalcol, color="MAFBRK")) + geom_abline()
  qqp <- qqp + geom_point() + theme_bw()
  qqp <- qqp + labs(x = "Expected -log(P.Value)", y = "Observed -log(P.Value)", color="MAF intervals") +
    theme(panel.border = element_blank(), legend.position = "right")
  qqp <- qqp + ggtitle(title) + scale_colour_brewer(palette="Blues")

  xannot <- max(df$EXP) - (max(df$EXP)*0.1)
  qqp <- qqp + annotate(
    "text",
    x = xannot,
    y = 0.5,
    label =
      paste0(
        "lambda: ",
        round(overall.lmb, 2)
      )
  )
  return(qqp)
}

pp <- myqqplot_mafbreak(
  df=regenieTable,
  pvalcol="LOG10P",
  log=TRUE,
  mafcol = "A1FREQ",
  title=paste("QQplot", params$phenotype, sep=" - ")
)

if(params$plot_ylimit == 0) {
  limit = NULL
} else {
  limit = c(0, params$plot_ylimit )
  pp <- pp + ylim(0, params$plot_ylimit)
  cat(paste("Please note that the Y-axis limit has been set to '", params$plot_ylimit,"'. Please use the 'plot_ylimit' parameter to adapt this behaviour."))
}
print(pp)
rm(regenieTable)
```

## Top Loci

```{r, echo=FALSE, message=FALSE, warning=FALSE}

annotatedTopHits <- fread(params$annotated_tophits_filename, select = c("CHROM", "GENPOS", "LOG10P", "ID", "GENE_NAME", "GENE_CHROMOSOME", "GENE_START", "GENE_END", "DISTANCE"), header=TRUE, tmpdir="$PWD")

topHitsGroupedByGene <-
  annotatedTopHits %>%
  group_by(GENE_NAME, GENE_CHROMOSOME, GENE_START, GENE_END) %>%
  summarise(GENE_LOG10P = max(LOG10P), VARIANTS = n()) %>%
  arrange(desc(GENE_LOG10P))

datatable(topHitsGroupedByGene)

```

## Validation and Logs

### Phenotype File Validation

```{r, echo=FALSE}
phenotype_log <- read.table(params$phenotype_log,
                      sep ="\t", header = TRUE, dec =".")
kable(phenotype_log)
```

### Covariate File Validation

```{r, echo=FALSE, results='asis'}
if(file.exists(params$covariate_log)) {
step1_log <- read.table(params$covariate_log,sep ="\t", header = TRUE, dec =".")
kable(step1_log)
} else {
    cat("*No covariates available.*")
}
```

### Regenie Step 1 Log

```{r, echo=FALSE, results='asis'}
if(file.exists(params$regenie_step1_log)) {
step1_log <- read.table(params$regenie_step1_log,sep ="\t", header = TRUE, dec =".")
kable(step1_log)
} else {
    cat("*Regenie step 1 skipped.*")
}
```

### Regenie Step 2 Log

```{r, echo=FALSE}
step2_log <- read.table(params$regenie_step2_log,
                      sep ="\t", header = TRUE, dec =".")
kable(step2_log)
```


---

<small>
This report has been created with **[nf-gwas `r params$version`](https://github.com/genepi/nf-gwas)**, a nextflow pipeline developed by [Sebastian Sch√∂nherr](mailto:sebastian.schoenherr@i-med.ac.at), [Lukas Forer](mailto:lukas.forer@i-med.ac.at) and [Johanna Schachtl-Riess](mailto:johanna.schachtl-riess@i-med.ac.at), Institute of Genetic Epidemiology, Medical University of Innsbruck.
</small>
